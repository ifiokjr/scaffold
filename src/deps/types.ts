export * from "https://cdn.skypack.dev/type-fest@2.13.1?dts";
