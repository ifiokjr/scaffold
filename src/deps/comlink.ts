export * from "https://esm.sh/comlink@4?target=deno";
