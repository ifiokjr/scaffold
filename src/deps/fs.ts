export * from "https://deno.land/std@0.145.0/fs/mod.ts";
