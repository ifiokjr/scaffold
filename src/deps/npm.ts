export { render } from "https://deno.land/x/eta@v1.12.3/mod.ts";

export * from "https://esm.sh/is-what@4.1.7?target=deno";
export * from "https://esm.sh/merge-anything@5.0.2?target=deno";
export { default as invariant } from "https://esm.sh/tiny-invariant@1.2.0?target=deno";
export * from "https://esm.sh/ts-extras@0.10.2?target=deno";
