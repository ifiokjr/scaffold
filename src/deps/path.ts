export { default as normalizePath } from "https://cdn.skypack.dev/normalize-path@v3.0.0?dts";
export * as path from "https://deno.land/std@0.144.0/path/mod.ts";
export * from "https://esm.sh/junk@4.0.0?type=deno";
