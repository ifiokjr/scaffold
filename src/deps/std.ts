export { Tar, Untar } from "https://deno.land/std@0.145.0/archive/tar.ts";
export { parse } from "https://deno.land/std@0.145.0/flags/mod.ts";
export { readLines } from "https://deno.land/std@0.145.0/io/mod.ts";
export {
  BaseHandler,
  type HandlerOptions,
} from "https://deno.land/std@0.145.0/log/handlers.ts";
export {
  type LevelName,
  LogLevels,
} from "https://deno.land/std@0.145.0/log/levels.ts";
export { Logger, LogRecord } from "https://deno.land/std@0.145.0/log/logger.ts";
export {
  globToRegExp,
  isGlob,
} from "https://deno.land/std@0.145.0/path/glob.ts";
export {
  copy,
  iterateReader,
  readableStreamFromIterable,
  readableStreamFromReader,
  readerFromIterable,
  readerFromStreamReader,
} from "https://deno.land/std@0.145.0/streams/conversion.ts";
export {
  assert,
  equal,
} from "https://deno.land/std@0.145.0/testing/asserts.ts";
