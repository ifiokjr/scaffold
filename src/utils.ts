import { colors, Spinner, type SpinnerOptions } from "./deps/cli.ts";
import { normalizePath, path } from "./deps/path.ts";

/**
 * Removes all undefined values from an object. Neither Firestore nor the
 * RealtimeDB allow `undefined` as a value.
 *
 * @param data The object to clean
 * @param nullish Set to true to clean null and undefined values
 */
export function removeUndefined<Shape extends object>(
  data: Shape,
  nullish = false,
) {
  const transformed = Object.create({});

  for (const [key, value] of Object.entries(data)) {
    if (nullish ? value == null : value === undefined) {
      continue;
    }

    transformed[key] = value;
  }

  return transformed;
}

const encoder = new TextEncoder();

/**
 * Encode a string for passing through to an anchor program.
 */
export function stringToUint8Array(text: string): Uint8Array {
  return encoder.encode(text);
}

const decoder = new TextDecoder();

/**
 * Decode a Uint8Array to a string.
 */
export function uint8ArrayToString(input: Uint8Array): string {
  return decoder.decode(input);
}

/**
 * Normalize a path to it's posix form.
 *
 * @param filePath The path to normalize
 * @param [stripTrailing=false] When set to true any trailing slashes will be
 * removed.
 *
 * @returns a normalized path with posix style separators
 */
export function normalize(filePath: string, stripTrailing?: boolean): string {
  return path.normalize(normalizePath(filePath, stripTrailing ?? false));
}

/**
 * Normalize a path as a directory with a trailing slash.
 *
 * @param directory The directory to normalize
 * @param [trailingSlash=true] When set to true a trailing slash will be added.
 */
export function normalizeDirectory(directory: string, trailingSlash = true) {
  return `${normalize(directory, true)}${trailingSlash ? "/" : ""}`;
}

/**
 * Get the normalized path of the provided `path`.
 */
export function getPath(filepath: string | URL): string {
  return filepath instanceof URL
    ? filepath.pathname
    : filepath.startsWith("file:")
    ? new URL(filepath).pathname
    : normalize(filepath);
}

interface ReadJsonOptions {
  /**
   * When set to true, the file with the default data will be created if it does
   * not exist.
   */
  create?: boolean;

  /**
   * The default data to use when the file does not exist.
   *
   * @default {}
   */
  defaultData?: object;
}

export async function readJson(path: string, options: ReadJsonOptions = {}) {
  const { create = false, defaultData = Object.create(null) } = options;

  try {
    return JSON.parse(await Deno.readTextFile(path));
  } catch {
    if (create) {
      await Deno.writeTextFile(path, JSON.stringify(defaultData));
    }

    return defaultData;
  }
}

export async function writeJson(path: string, json: object) {
  try {
    const content = JSON.stringify(json);
    await Deno.writeTextFile(path, content, { create: true });
  } catch {
    return {};
  }
}

export function wait(options: SpinnerOptions) {
  let spinner: Spinner;
  const getSpinner = () => {
    return spinner ??= new Spinner({
      text: options.text,
      prefix: options.prefix ?? "",
      color: options.color ?? colors.cyan,
      spinner: options.spinner ?? "dots",
      hideCursor: options.hideCursor ?? true,
      indent: options.indent ?? 0,
      interval: options.interval ?? 100,
      stream: options.stream ?? Deno.stdout,
      enabled: true,
      discardStdin: true,
    }).start();
  };

  const proxy = {
    text: (text: string) => {
      if (options.enabled === false) return;
      getSpinner().text = text;
      return proxy;
    },
    info: (text: string) => {
      if (options.enabled === false) return;
      getSpinner().info(text);
      return proxy;
    },
    fail: (text: string) => {
      if (options.enabled === false) return;
      getSpinner().fail(text);
      return proxy;
    },
    succeed: (text: string) => {
      if (options.enabled === false) return;
      getSpinner().succeed(text);
      return proxy;
    },
    warn: (text: string) => {
      if (options.enabled === false) return;
      getSpinner().warn(text);
      return proxy;
    },
    stop: () => {
      if (options.enabled === false) return;
      getSpinner().stop();
      return proxy;
    },
  };

  return proxy;
}
