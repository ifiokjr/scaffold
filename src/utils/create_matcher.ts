import { isFunction, isString } from "../deps/npm.ts";
import { isJunk, path } from "../deps/path.ts";
import { globToRegExp, isGlob } from "../deps/std.ts";

export type MatchFunction = (source: string) => boolean;
export type Match = string | MatchFunction | RegExp;

export interface CreateMatcherOptions {
  /**
   * When a glob begins with `!` and the glob matches the source, the match will
   * be negated. Set this to `true` to disable the behavior.
   *
   * @default false
   */
  disableNegation?: boolean;

  /**
   * Expand the glob when it ends with `/`.
   *
   * @default false
   */
  expandGlobs?: boolean;

  /**
   * Set to true to make globs case insensitive.
   *
   * @default false
   */
  caseInsensitive?: boolean;

  /**
   * When true will match dot files and folders.
   *
   * @default false
   */
  dot?: boolean;

  /**
   * When true will match junk files
   *
   * @default false
   */
  junk?: boolean;

  /**
   * The extensions that can be matched. Setting this to anything other than an
   * empty array / or null will prevent matches on directories. Unless the
   * directory has an extension 🤷‍♀️.
   *
   * @default undefined
   */
  extensions?: string[] | null;
}
/**
 * Transform a pattern into a matching function.
 */

export function createMatcher(
  pattern: Match | Match[],
  options: CreateMatcherOptions = {},
): MatchFunction {
  const { disableNegation = false, expandGlobs = false } = options;
  const matchers = Array.isArray(pattern) ? pattern : [pattern].filter(Boolean);

  return (source) => {
    if (
      // Exit early if no matchers are provided.
      matchers.length === 0 ||
      // Exit early if the source is a dot file / folder.
      (
        !options.dot &&
        source.split(path.sep).some((part) => part.startsWith("."))
      ) ||
      // Exit early if the source is a junk file / folder.
      (
        !options.junk && (isJunk(source) || isJunk(path.basename(source)))
      ) ||
      // Exit early if not a valid extension.
      (
        options.extensions &&
        !options.extensions.some((extension) => source.endsWith(extension))
      )
    ) {
      return false;
    }

    let anyMatch = false;

    for (const matcher of matchers) {
      let match = false;
      // True when this is a string glob starting with `!`.
      let negated = false;

      if (isFunction(matcher)) {
        match = matcher(source) || match;
      } else if (isString(matcher) && !isGlob(matcher)) {
        match = matcher === source || match;
      } else {
        let regex: RegExp;

        if (isString(matcher)) {
          negated = disableNegation || matcher[0] === "!";
          regex = globToRegExp(
            `${negated ? matcher.slice(1) : matcher}${
              matcher.endsWith("/") && expandGlobs ? "**/*" : ""
            }`,
            { caseInsensitive: options.caseInsensitive },
          );
        } else {
          regex = matcher;
        }

        match = regex.test(source) || match;
      }

      if (match && disableNegation) {
        // exit early when a match is found and if negated return false.
        return true;
      }

      if (match && negated) {
        return false;
      }
      anyMatch = match || anyMatch;
    }

    return anyMatch;
  };
}
