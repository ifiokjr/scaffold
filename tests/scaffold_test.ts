import { uint8ArrayToString } from "../src/utils.ts";
import { assertSnapshot, describe, it } from "./deps.ts";
import { run } from "./helpers.ts";

describe("scaffold", () => {
  it("can display help", async (t) => {
    const command = run(["--help"], { stdout: "piped" });
    const output = uint8ArrayToString(await command.output());
    command.close();

    await assertSnapshot(t, output);
  });
});
