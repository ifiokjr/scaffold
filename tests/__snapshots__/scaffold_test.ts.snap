export const snapshot = {};

snapshot[`global > scaffold > can display help 1`] = `
'
  \\x1b[1mUsage:\\x1b[22m   \\x1b[35mscaffold \\x1b[33m<\\x1b[35m\\x1b[35mrepo\\x1b[35m\\x1b[33m>\\x1b[35m \\x1b[33m[\\x1b[35m\\x1b[35mfolder\\x1b[35m\\x1b[33m]\\x1b[35m\\x1b[39m
  \\x1b[1mVersion:\\x1b[22m \\x1b[33m0.1.0\\x1b[39m                   

  \\x1b[1mDescription:\\x1b[22m

    🏗️ Scaffold a new project from any GitHub, GitLab or BitBucket git repository.

  \\x1b[1mOptions:\\x1b[22m

    \\x1b[34m-h\\x1b[39m, \\x1b[34m--help\\x1b[39m                     \\x1b[31m\\x1b[1m-\\x1b[22m\\x1b[39m Show this help.                                                                                                                    
    \\x1b[34m-V\\x1b[39m, \\x1b[34m--version\\x1b[39m                  \\x1b[31m\\x1b[1m-\\x1b[22m\\x1b[39m Show the version number for this program.                                                                                          
    \\x1b[34m-c\\x1b[39m, \\x1b[34m--cache\\x1b[39m           \\x1b[33m[\\x1b[39m\\x1b[35mcache\\x1b[39m\\x1b[33m]\\x1b[39m  \\x1b[31m\\x1b[1m-\\x1b[22m\\x1b[39m Set the cache directory relative to the current directory.                                                                         
    \\x1b[34m--no-cache\\x1b[39m                     \\x1b[31m\\x1b[1m-\\x1b[22m\\x1b[39m Disable the cache.                                                                                                                 
    \\x1b[34m-d\\x1b[39m, \\x1b[34m--debug\\x1b[39m                    \\x1b[31m\\x1b[1m-\\x1b[22m\\x1b[39m Enable debug logging (shorthand for \\x1b[3m\\x1b[90m--log-level=debug\\x1b[39m\\x1b[23m                                                                              
    \\x1b[34m-f\\x1b[39m, \\x1b[34m--force\\x1b[39m                    \\x1b[31m\\x1b[1m-\\x1b[22m\\x1b[39m Overwrite files even if they already exist.                                                                                        
    \\x1b[34m-l\\x1b[39m, \\x1b[34m--log-level\\x1b[39m       \\x1b[33m[\\x1b[39m\\x1b[35mlevel\\x1b[39m\\x1b[33m]\\x1b[39m  \\x1b[31m\\x1b[1m-\\x1b[22m\\x1b[39m Set the log level.                                                              (\\x1b[1mValues: \\x1b[22m\\x1b[32m"debug"\\x1b[39m, \\x1b[32m"info"\\x1b[39m, \\x1b[32m"warn"\\x1b[39m, \\x1b[32m"error"\\x1b[39m, \\x1b[32m"fatal"\\x1b[39m)
    \\x1b[34m-s\\x1b[39m, \\x1b[34m--silent\\x1b[39m                   \\x1b[31m\\x1b[1m-\\x1b[22m\\x1b[39m Disable all logging.                                                                                                               
    \\x1b[34m--no-template\\x1b[39m                  \\x1b[31m\\x1b[1m-\\x1b[22m\\x1b[39m Disable loading the \\x1b[3m\\x1b[90mtemplate.config.ts\\x1b[39m\\x1b[23m file.                                                                                       
    \\x1b[34m-n\\x1b[39m, \\x1b[34m--name\\x1b[39m            \\x1b[33m[\\x1b[39m\\x1b[35mname\\x1b[39m\\x1b[33m]\\x1b[39m   \\x1b[31m\\x1b[1m-\\x1b[22m\\x1b[39m Set the name to be used in the template                                                                                            
    \\x1b[34m--description\\x1b[39m         \\x1b[33m[\\x1b[39m\\x1b[35mname\\x1b[39m\\x1b[33m]\\x1b[39m   \\x1b[31m\\x1b[1m-\\x1b[22m\\x1b[39m Set the description to be used in the template                                                                                     
    \\x1b[34m-y\\x1b[39m, \\x1b[34m--no-interactive\\x1b[39m           \\x1b[31m\\x1b[1m-\\x1b[22m\\x1b[39m Disable the interactive prompt.                                                                                                    
    \\x1b[34m--use-temp-source\\x1b[39m              \\x1b[31m\\x1b[1m-\\x1b[22m\\x1b[39m Copy local files to a temporary directory before scaffolding. This might break                                                     
                                     local imports.                                                                                                                     

'
`;

snapshot[`global > scaffold > can create templates from local folders 1`] = `
{
  ".gitignore": "NOTHING",
  ".should-include": "",
  "asdf/index.js": 'import * as fs from "node:fs/promises";

export function asdf(filepath) {
  return fs.writeFile(filepath, "asdf");
}
',
  "asdf/package.json": '{
  "name": "asdf",
  "version": "1.0.0",
  "type": "module",
  "exports": "./index.js"
}
',
  file: "",
  "hello.text": "hello world",
  "package.json": '{
  "name": "scaffold",
  "description": "A scaffold for Deno",
  "dependencies": {
    "asdf": "^1.2.2"
  },
  "version": "1.0.0"
}
',
  "rename-thee.md": "Should be renamed from \`rename-me.md\` to \`rename-thee.md\`.
",
  "scaffold.json": "{}
",
}
`;
